File: JointResolver/RigidAnalyzer/RotationalJoint.vb
Imports Inventor

Module RotationalJoint
    Public Function isRotationalJoint(ByRef jointI As CustomRigidJoint) As Boolean
        If jointI.joints.Count = 1 Then
            Dim joint As AssemblyJointDefinition = jointI.joints.Item(0).Definition
            Return joint.JointType = AssemblyJointTypeEnum.kRotationalJointType OrElse _
                (joint.JointType = AssemblyJointTypeEnum.kCylindricalJointType AndAlso _
                 (Not (joint.HasAngularPositionLimits) OrElse joint.AngularPositionEndLimit.Value <> joint.AngularPositionStartLimit.Value))
        End If
        Return False
    End Function

    Public Function getInfo(ByRef jointI As CustomRigidJoint, ByRef parent As CustomRigidGroup)
        If Not (isRotationalJoint(jointI)) Then Return Nothing

        Dim joint As AssemblyJointDefinition = jointI.joints.Item(0).Definition
        Dim child As CustomRigidGroup = Nothing
        Dim first As Boolean = False
        If (jointI.groupOne.Equals(parent)) Then
            child = jointI.groupTwo
            first = True
        ElseIf jointI.groupTwo.Equals(parent) Then
            child = jointI.groupOne
        End If
        If IsNothing(child) Then Return Nothing

        Dim groupANormal, groupBNormal As UnitVector
        Dim groupABase, groupBBase As Point

        If (joint.JointType = AssemblyJointTypeEnum.kCylindricalJointType) Then
            groupANormal = joint.OriginOne.Geometry.Geometry.AxisVector
            groupABase = joint.OriginOne.Geometry.Geometry.BasePoint
            groupBNormal = joint.OriginTwo.Geometry.Geometry.AxisVector
            groupBBase = joint.OriginTwo.Geometry.Geometry.BasePoint
        Else
            groupANormal = joint.OriginOne.Geometry.Geometry.Normal
            groupABase = joint.OriginOne.Geometry.Geometry.Center
            groupBNormal = joint.OriginTwo.Geometry.Geometry.Normal
            groupBBase = joint.OriginTwo.Geometry.Geometry.Center
        End If

        Dim currentAngularPosition As Double = If(Not (IsNothing(joint.AngularPosition)), joint.AngularPosition.Value, 0)
        Dim hasAngularLimit As Boolean = joint.HasAngularPositionLimits
        Dim angularLimitLow, angularLimitHigh As Double
        If (hasAngularLimit) Then
            angularLimitLow = joint.AngularPositionStartLimit.Value
            angularLimitHigh = joint.AngularPositionEndLimit.Value
        End If
        Return Main.printVector(If(first, groupABase, groupBBase)) & ":" & Main.printVector(If(first, groupANormal, groupBNormal)) & ":" & Main.printVector(If(first, groupBBase, groupABase)) & ":" & Main.printVector(If(first, groupBNormal, groupANormal))
    End Function
End Module

File: JointResolver-Rev2/SkeletalStructure/RotationalJoint.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using Inventor;

public class RotationalJoint : RotationalJoint_Base, InventorSkeletalJoint
{
    private SkeletalJoint wrapped;

    public SkeletalJoint getWrapped() { return wrapped; }

    public static bool isRotationalJoint(CustomRigidJoint jointI)
    {
        if (jointI.joints.Count == 1)
        {
            AssemblyJointDefinition joint = jointI.joints[0].Definition;
            //Checks if there is no linear motion allowed.
            return joint.JointType == AssemblyJointTypeEnum.kRotationalJointType
                || (joint.JointType == AssemblyJointTypeEnum.kCylindricalJointType
                && joint.HasLinearPositionEndLimit && joint.HasLinearPositionEndLimit
                && joint.LinearPositionStartLimit() == joint.LinearPositionEndLimit());
        }
        return false;
    }

    private static void getRotationalInfo(dynamic geom, out UnitVector groupANormal, out Point groupABase)
    {
        if (geom is EdgeProxy)
        {
            EdgeProxy edge = (EdgeProxy)geom;
            if (edge.GeometryType == CurveTypeEnum.kCircularArcCurve || edge.GeometryType == CurveTypeEnum.kCircleCurve)
            {
                groupANormal = geom.Geometry.Normal;
                groupABase = geom.Geometry.Center;
            }
            else if (edge.GeometryType == CurveTypeEnum.kLineSegmentCurve || edge.GeometryType == CurveTypeEnum.kLineCurve)
            {
                groupANormal = edge.Geometry.Direction;
                groupABase = edge.Geometry.MidPoint;
                // mid points look right...
            }
            else
            {
                throw new Exception("Unimplemented " + Enum.GetName(typeof(CurveTypeEnum), edge.GeometryType));
            }
        }
        else if (geom is FaceProxy)
        {
            FaceProxy face = (FaceProxy)geom;
            Console.WriteLine("FaceType: " + Enum.GetName(typeof(SurfaceTypeEnum), face.SurfaceType));
            if (face.SurfaceType == SurfaceTypeEnum.kPlaneSurface)
            {
                groupANormal = face.Geometry.Normal;
                groupABase = face.Geometry.RootPoint;
            }
            else if (face.SurfaceType == SurfaceTypeEnum.kCylinderSurface)
            {
                groupABase = face.Geometry.BasePoint;
                groupANormal = face.Geometry.AxisVector;
                // face.Geometry.Radius
            }
            else
            {
                throw new Exception("Unimplemented surface type " + Enum.GetName(typeof(SurfaceTypeEnum), face.SurfaceType));
            }
        }
        else
        {
            throw new Exception("Unimplemented proxy object " + Enum.GetName(typeof(ObjectTypeEnum), geom.Type));
        }
    }

    public RotationalJoint(CustomRigidGroup parent, CustomRigidJoint rigidJoint)
    {
        if (!(isRotationalJoint(rigidJoint)))
            throw new Exception("Not a rotational joint");
        wrapped = new SkeletalJoint(parent, rigidJoint);

        UnitVector groupANormal = null;
        UnitVector groupBNormal = null;
        Point groupABase = null;
        Point groupBBase = null;
        Console.WriteLine("O1: " + Enum.GetName(typeof(ObjectTypeEnum), wrapped.asmJoint.OriginOne.Type) + "\t" +
            Enum.GetName(typeof(ObjectTypeEnum), wrapped.asmJoint.OriginOne.Geometry.Type));
        Console.WriteLine("O2: " + Enum.GetName(typeof(ObjectTypeEnum), wrapped.asmJoint.OriginTwo.Type) + "\t" +
            Enum.GetName(typeof(ObjectTypeEnum), wrapped.asmJoint.OriginTwo.Geometry.Type));


        getRotationalInfo(wrapped.asmJoint.OriginOne.Geometry, out groupANormal, out groupABase);
        getRotationalInfo(wrapped.asmJoint.OriginTwo.Geometry, out groupBNormal, out groupBBase);

        if (wrapped.childIsTheOne)
        {
            childNormal = Utilities.toBXDVector(groupANormal);
            childBase = Utilities.toBXDVector(groupABase);
            parentNormal = Utilities.toBXDVector(groupBNormal);
            parentBase = Utilities.toBXDVector(groupBBase);
        }
        else
        {
            childNormal = Utilities.toBXDVector(groupBNormal);
            childBase = Utilities.toBXDVector(groupBBase);
            parentNormal = Utilities.toBXDVector(groupANormal);
            parentBase = Utilities.toBXDVector(groupABase);
        }

        currentAngularPosition = !((wrapped.asmJoint.AngularPosition == null)) ? (float) wrapped.asmJoint.AngularPosition.Value : 0;
        hasAngularLimit = wrapped.asmJoint.HasAngularPositionLimits;
        if ((hasAngularLimit))
        {
            angularLimitLow = (float) wrapped.asmJoint.AngularPositionStartLimit.Value;
            angularLimitHigh = (float) wrapped.asmJoint.AngularPositionEndLimit.Value;
        }
    }

    protected override string ToString_Internal()
    {
        return wrapped.childGroup + " rotates about " + wrapped.parentGroup;
    }
}